### PR TITLE
[WIP] Notes about closure de-sugaring

### PR DESCRIPTION
**Pull request by blitzerr**

### CODE REVIEW

File: src/closure.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Closure Expansion in rustc
+
+Let's start with an example
+
+### Example 1
+```rust
```
Comment (nikomatsakis) [inline src/closure.md]: A little introduction here too perhaps. Maybe

---

To start, let's take a look at how the closure in the following example is desugared:

File: src/closure.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+fn foo<F>(closure: F)
+  where F: Fn() {
+  closure()
+}
+fn main() {
+  let x = 10;
+  foo(|| println!("Hi {}", x))
+}
+```
+If you compile Example 1 using the command
```
Comment (nikomatsakis) [inline src/closure.md]: "To see how the closure gets desugared, we'll take a look at the MIR that gets generated. To generate the MIR, we need to run `rustc` with the `-Zdump-mir=all` argument, like so:"

(We can link the MIR references to the MIR chapter; not sure if there is a chapter that describes dump-mir...)

File: src/closure.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+```
+rustc +stage1 test.rs -Zdump-mir=all
+```
+where the name of the file is test.rs. This command will generate a directory called mir_dump
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/closure.md]: "After we run this command, rustc will generate a directory called `mir_dump` ..."

File: src/closure.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+which will contain several files. If we look at file `rustc.main.-------.mir_map.0.mir`, we will find this
+
+```rust,ignore
+  _3 = [closure@test.rs:9:9: 11:6] { x: move _4 }; // bb0[6]: scope 1 at test.rs:9:9: 11:6
+                                   // closure
```
Comment (nikomatsakis) [inline src/closure.md]: I'd probably cut all the `//` stuff and just focus on the MIR line itself

File: src/closure.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                                   // + def_id: DefId(0/1:10 ~ test[317d]::main[0]::{{closure}}[0])
+                                   // + substs: ClosureSubsts {
+                                   //     substs: [
+                                   //         i8,
+                                   //         extern "rust-call" fn(()),
+                                   //         &i32
+                                   //     ]
+                                   // }
+```
```
Comment (nikomatsakis) [inline src/closure.md]: Here, I think we should say a bit more =)

This is showing the construction of the desugared closure struct. The `[closure@test.rs...]` type is the "name" of the closure struct, and the `{ x : move _4 }` shows the value for each of its fields. In this case, we can see that we are *moving* the variable `_4` into the closure struct. If we look back to see where `_4` is defined ...

(I left the last sentence finished: what do we find if we do that? I presume it is a borrow of `x`...

**Comment by mark-i-m** [comment]
@blitzerr Ping. Any updates?

**Comment by mark-i-m** [comment]
Ping @blitzerr Any updates?

**Comment by mark-i-m** [comment]
@nikomatsakis Ping :)

**Comment by mark-i-m** [comment]
@nikomatsakis Ping

**Comment by mark-i-m** [comment]
@nikomatsakis Ping? Do you have further thoughts on this PR?

**Comment by mark-i-m** [comment]
@blitzerr Thanks for your patience. It appears that Niko is busy, and I finally got around to reading this PR myself. I opened PR https://github.com/blitzerr/rustc-guide/pull/1 on your repo with some changes I propose. Please let me know what you think. If they looks good to you go ahead and merge that PR so that those commits get added here. After that, I think we can probably just go ahead and merge it and update from there.

Let me know if you have any questions.

**Comment by blitzerr** [comment]
That sounds good. I will take a look tonight.

**Comment by mark-i-m** [comment]
Ping @blitzerr. Will you get a chance to look soon at my PR?

**Comment by mark-i-m** [comment]
Ping @blitzerr. I hope you are doing well. Will you be able to look at this?

**Comment by blitzerr** [comment]
Hi,

I am sorry for long wait. I will take a look this weekend

**Comment by mark-i-m** [comment]
Friendly ping @blitzerr  :ping_pong:

**Pull request closed without merging**